### PR TITLE
Implement service-worker registration and caching

### PR DESCRIPTION
**Pull request by jamesbhobbs**
Implements `offline-plugin` for SW registration and caching.

Outstanding to be done in another PR:
- Offline support
- Implementation of react router client-side updates (see inline comments for minor work 
- Redirection of HTTP to HTTPS and local certs needs to be done in a separate PR

* [x] Fixes https://github.com/bbc/simorgh/issues/72
* [x] Tests added for new features
* [ ] ~~E2E tests added for new features~~ This is very difficult.

Testing can be run locally using `npm run test:lighthouse` and in future this will be auto-tested by https://github.com/bbc/simorgh/issues/331

* [ ] Test engineer approval

image

### CODE REVIEW

File: public/manifest.json [hunk 1]
```diff
@@ -1,11 +1,11 @@
 {
-  "name": "Simorgh",
-  "short_name": "Simorgh",
+  "name": "BBC News",
+  "short_name": "BBC News",
   "Scope": "/",
-  "background_color": "#bb1919",
-  "display": "browser",
+  "background_color": "#b80000",
+  "display": "standalone",
```
Comment (jamesbhobbs) [inline public/manifest.json]: Purely done to pass the lighthouse test, why doesn't it allow 'browser'?

File: razzle.config.js [hunk 2]
```diff
@@ -1,7 +1,29 @@
...
+      appConfig.plugins.push(
+        new OfflinePlugin({
+          appShell: '/article/',
+          autoUpdate: 1000 * 60 * 2,
+          // externals means files not generated by webpack
+          // an alternative would be html-webpack-plugin
+          externals: ['/'],
+          ServiceWorker: {
+            events: true,
+            // obviously we need a file to match this path
```
Comment (jamesbhobbs) [inline razzle.config.js]: Will do a hello world style file in this PR

File: src/client.js [hunk 3]
```diff
@@ -4,8 +4,16 @@ import 'babel-polyfill';
 import { hydrate } from 'react-dom';
 import { BrowserRouter } from 'react-router-dom';
 import { ensureReady, After } from '@jaredpalmer/after';
+import * as OfflinePluginRuntime from 'offline-plugin/runtime';
 import routes from './app/routes';
 
+OfflinePluginRuntime.install({
+  // Source: https://zach.codes/handling-client-side-app-updates-with-service-workers/
+  // NB this means links must have special logic. I have no links to add this on yet, see the link above.
```
Comment (jamesbhobbs) [inline src/client.js]: Suggest this isn't done in this PR

**Comment by benjaminhobbs** [comment]
Broken build is conflicting with a hack we have in place for the eslint use in production. Need to debug the razzle config.

File: razzle.config.js [hunk 4]
```diff
@@ -54,6 +57,23 @@ module.exports = {
       };
     }
 
+    if (target === 'web') {
+      appConfig.plugins.push(
+        new OfflinePlugin({
+          appShell: '/article/',
```
Comment (jtart) [inline razzle.config.js]: Have you seen this? https://github.com/NekR/offline-plugin/blob/abdae89fd777ed7e6ea05ad31c993ffaf3593b5b/docs/app-shell.md#ssr. I would potentially recommend removing this config value for now.

File: razzle.config.js [hunk 5]
```diff
@@ -54,6 +57,23 @@ module.exports = {
       };
     }
 
+    if (target === 'web') {
```
Comment (jtart) [inline razzle.config.js]: Service worker setup is failing for me on local dev. Recommend to enable just for prod build.

File: razzle.config.js [hunk 6]
```diff
@@ -54,6 +57,23 @@ module.exports = {
...
+      appConfig.plugins.push(
+        new OfflinePlugin({
+          appShell: '/article/',
+          autoUpdate: 1000 * 60 * 2,
```
Comment (jtart) [inline razzle.config.js]: Setting this to 2 minutes seems awfully low, unless you are considering PWA still under development and want faster updates.

Also, please add a comment that this is in milliseconds.

File: razzle.config.js [hunk 7]
```diff
@@ -54,6 +57,23 @@ module.exports = {
...
+          // externals means files not generated by webpack
+          // an alternative would be html-webpack-plugin
+          externals: ['/'],
+          ServiceWorker: {
+            events: true,
+            navigateFallbackURL: './offline.html',
```
Comment (jtart) [inline razzle.config.js]: This should link to`/offline.html`

File: src/client.js [hunk 8]
```diff
@@ -4,8 +4,16 @@ import 'babel-polyfill';
 import { hydrate } from 'react-dom';
 import { BrowserRouter } from 'react-router-dom';
 import { ensureReady, After } from '@jaredpalmer/after';
+import * as OfflinePluginRuntime from 'offline-plugin/runtime';
 import routes from './app/routes';
 
+OfflinePluginRuntime.install({
```
Comment (jtart) [inline src/client.js]: Please add tests for this.

Also, if we are only running the server worker on production, it should be wrapped in a conditional.

**Review by jtart — DISMISSED** [review]
I think `offline-plugin` is fine in the short-term to get a PWA off the ground but I am not a big fan of it.

Unless I am missing something, it does not provide a flexible caching solution at all really, there's only [2 caching strategies](https://github.com/NekR/offline-plugin/blob/master/docs/options.md#responsestrategy-cache-first--network-first), and very limited, if any, scope to config caching strategies across different URL globs.

While we haven't defined caching strategies, I would not want to tie our hands by using a solution that is not flexible enough. I would question whether Workbox is more complex to setup, and as we are not really in a rush to get this through, it would be good to investigate Workbox closer and a solution similar to this implemented. We could then do a comparison of the two and decide on the best.

I have submitted a review of the implementation, in case it is decided that we should merge in this work until a Workbox investigation goes ahead.

File: razzle.config.js (hunk 6 above)
Comment (jamesbhobbs) [reply razzle.config.js]: Also look into divergent caching strategies for different assets, is it possible and is there an easy way to do it?

File: razzle.config.js (hunk 7 above)
Comment (jamesbhobbs) [reply razzle.config.js]: Done

File: razzle.config.js (hunk 5 above)
Comment (jamesbhobbs) [reply razzle.config.js]: Done

File: src/client.js (hunk 8 above)
Comment (jamesbhobbs) [reply src/client.js]: Done 2nd half

**Comment by sareh** [comment]
Looks great. 92 score in Lighthouse as well.  
screen shot of lighthouse audit output with 92 score

File: src/client.js [hunk 9]
```diff
@@ -4,8 +4,18 @@ import 'babel-polyfill';
 import { hydrate } from 'react-dom';
 import { BrowserRouter } from 'react-router-dom';
 import { ensureReady, After } from '@jaredpalmer/after';
+import * as OfflinePluginRuntime from 'offline-plugin/runtime';
 import routes from './app/routes';
 
+if (process.env.NODE_ENV === 'production') {
+  OfflinePluginRuntime.install({
+    // Source: https://zach.codes/handling-client-side-app-updates-with-service-workers/
+    // NB this means links must have special logic. I have no links to add this on yet, see the link above.
```
Comment (sareh) [inline src/client.js:13]: We can use the Link here: https://github.com/bbc/simorgh/blob/latest/src/app/containers/InlineLink/index.jsx#L6

File: razzle.config.js [hunk 10]
```diff
@@ -28,10 +30,29 @@ module.exports = {
             },
           },
         };
+
+        // Setup service worker and offline support
+        appConfig.plugins.push(
+          new OfflinePlugin({
+            AppCache: false, // because it's deprecated
```
Comment (pjlee11) [inline razzle.config.js:37]: Good comment - could you link on the PR where into documented that this is deprecated?

File: src/client.js (hunk 9 above)
Comment (sareh) [reply src/client.js:13]: If updating the SW on each Link onClick is not going to be implemented in this PR, I'd suggest opening an issue to capture this work, so it is not forgotten.

**Review by sareh — DISMISSED** [review]

File: src/client.js (hunk 9 above)
Comment (jtart) [reply src/client.js:13]: Issue for SW busting https://github.com/bbc/simorgh/issues/571

File: razzle.config.js (hunk 10 above)
Comment (jtart) [reply razzle.config.js:37]: https://developer.mozilla.org/en-US/docs/Web/HTML/Using_the_application_cache

File: public/manifest.json [hunk 11]
```diff
@@ -1,10 +1,10 @@
 {
   "name": "BBC News",
   "short_name": "BBC News",
-  "Scope": "/",
+  "Scope": "/news/",
   "background_color": "#b80000",
   "display": "standalone",
-  "start_url": "/",
+  "start_url": "/news/",
```
Comment (pjlee11) [inline public/manifest.json:7]: With this being hardcoded to a service such as `/news/` will we need a manifest per service once the WS sites are included?

**Review by pjlee11 — DISMISSED** [review]
Group reviewed by the team

**Comment by jamesbrumpton** [comment]
LGTM

**Pull request merged**